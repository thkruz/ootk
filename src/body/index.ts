export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
