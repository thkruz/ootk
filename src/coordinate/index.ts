export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
