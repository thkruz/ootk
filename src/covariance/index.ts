export * from './CovarianceSample.js';
export * from './StateCovariance.js';
