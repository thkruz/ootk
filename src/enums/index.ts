export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
