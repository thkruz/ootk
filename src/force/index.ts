export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

