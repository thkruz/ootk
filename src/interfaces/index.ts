export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
