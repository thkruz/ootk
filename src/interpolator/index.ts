export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

