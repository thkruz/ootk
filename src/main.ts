/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
