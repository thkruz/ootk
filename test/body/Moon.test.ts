import { Degrees, EpochUTC, Moon, Vector3D } from '../../src/main';

describe('Moon', () => {
  let exampleDate: Date;

  beforeEach(() => {
    const mockExampleDate = new Date(1705109326817);

    exampleDate = new Date(mockExampleDate.getTime());
  });

  // The static property 'mu' should be accessible and have a value of 4902.799.
  it('should have a static property "mu" with value 4902.799', () => {
    expect(Moon.mu).toBe(4902.799);
  });

  /*
   * The static property 'radiusEquator' should be accessible and have a value
   * of 1738.0.
   */
  it('should have a static property "radiusEquator" with value 1738.0', () => {
    expect(Moon.radiusEquator).toBe(1738.0);
  });

  /*
   * The static method 'eci' should be callable with an "EpochUTC" parameter and
   * return a 'Vector3D' object.
   */
  it('should be callable with an "EpochUTC" parameter and return a "Vector3D" object', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const result = Moon.eci(epoch);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'illumination' should be callable with an "EpochUTC"
   * parameter and an optional 'origin' parameter, and return a number between 0
   * and 1.
   */
  it('should be callable with an "EpochUTC" parameter and an optional "origin" parameter', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const result = Moon.illumination(epoch);

    expect(result).toBeGreaterThanOrEqual(0);
    expect(result).toBeLessThanOrEqual(1);
  });

  /*
   * The static method 'diameter' should be callable with two 'Vector3D'
   * parameters and return a number.
   */
  it('should be callable with two "Vector3D" parameters and return a number', () => {
    const obsPos = new Vector3D(1, 2, 3);
    const moonPos = new Vector3D(4, 5, 6);
    const result = Moon.diameter(obsPos, moonPos);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'eci' should throw an error if the "EpochUTC" parameter
   * is not provided.
   */
  it('should throw an error if the "EpochUTC" parameter is not provided', () => {
    expect(() => {
      Moon.eci();
    }).not.toThrow('EpochUTC parameter is required');
  });

  /*
   * The static method 'illumination' should return 0.5 if the 'origin'
   * parameter is not provided.
   */
  it('should return 0.5 if the "origin" parameter is not provided', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const result = Moon.illumination(epoch);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'diameter' should return 0 if the two 'Vector3D'
   * parameters are the same.
   */
  it('should return NaN if the two "Vector3D" parameters are the same', () => {
    const pos = new Vector3D(1, 2, 3);
    const result = Moon.diameter(pos, pos);

    expect(result).toBeNaN();
  });

  /*
   * The static method 'getMoonIllumination' should be callable with a 'number'
   * or 'Date' parameter and return a 'MoonIlluminationData' object.
   */
  it('should return a MoonIlluminationData object when called with a number parameter', () => {
    const date = 1635724800000; // November 1, 2021
    const result = Moon.getMoonIllumination(date);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'rae' should be callable with a 'Date', 'Degrees', and
   * 'Degrees' parameters and return an object with 'az', 'el', 'rng', and
   * 'parallacticAngle' properties.
   */
  it('should return an object with \'az\', \'el\', \'rng\', and \'parallacticAngle\' properties', () => {
    const date = new Date(1635724800000); // November 1, 2021
    const lat = 37.7749 as Degrees; // San Francisco latitude
    const lon = -122.4194 as Degrees; // San Francisco longitude
    const result = Moon.rae(date, lat, lon);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'getMoonTimes' should be callable with a 'Date',
   * 'Degrees', 'Degrees', and 'boolean' parameters and return an object with
   * 'rise', 'set', 'ye', 'alwaysUp', 'alwaysDown', and 'highest' properties.
   */
  it('should return with \'rise\', \'set\', \'ye\', \'alwaysUp\', \'alwaysDown\', \'highest\' properties', () => {
    const date = exampleDate;
    const lat = 37.7749 as Degrees; // San Francisco latitude
    const lon = -122.4194 as Degrees; // San Francisco longitude
    const isUtc = false;
    const result = Moon.getMoonTimes(date, lat, lon, isUtc);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'rae' should be callable with a 'Date', 'Degrees', and
   * 'Degrees' parameters and return an object with 'az', 'el', 'rng', and
   * 'parallacticAngle' properties.
   */
  it('should return an object with \'az\', \'el\', \'rng\', and \'parallacticAngle\' properties', () => {
    const date = new Date(exampleDate); // November 1, 2021
    const lat = 37.7749 as Degrees; // San Francisco latitude
    const lon = -122.4194 as Degrees; // San Francisco longitude
    const result = Moon.rae(date, lat, lon);

    expect(result).toMatchSnapshot();
  });

  /*
   * The static method 'getMoonTimes' should be callable with a 'Date',
   * 'Degrees', 'Degrees', and 'boolean' parameters and return an object with
   * 'rise', 'set', 'ye', 'alwaysUp', 'alwaysDown', and 'highest' properties.
   */
  it('should return with \'rise\', \'set\', \'ye\', \'alwaysUp\', \'alwaysDown\', and \'highest\' properties', () => {
    const date = new Date(exampleDate.getTime() - exampleDate.getTimezoneOffset() * 60000);
    const lat = 37.7749 as Degrees; // San Francisco latitude
    const lon = -122.4194 as Degrees; // San Francisco longitude
    const isUtc = false;
    const result = Moon.getMoonTimes(date, lat, lon, isUtc);

    expect(result).toMatchSnapshot();
  });

  /*
   * The private method 'moonCoords' should be callable with a 'number'
   * parameter and return a 'RaDec' object.
   */
  it('should call the private method \'moonCoords\' with a number parameter and return a RaDec object', () => {
    const d = 123456789;
    const result = Moon.moonCoords(d);

    expect(result).toMatchSnapshot();
  });
});
