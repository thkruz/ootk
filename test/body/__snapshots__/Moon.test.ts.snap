// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Moon should be callable with an "EpochUTC" parameter and return a "Vector3D" object 1`] = `
Vector3D {
  "x": 248471.54051777,
  "y": -229511.18274819673,
  "z": -131233.28802621315,
}
`;

exports[`Moon should be callable with two "Vector3D" parameters and return a number 1`] = `NaN`;

exports[`Moon should call the private method 'moonCoords' with a number parameter and return a RaDec object 1`] = `
Object {
  "dec": -0.21428784368024154,
  "dist": 401235.63978571084,
  "ra": -0.7625969983284882,
}
`;

exports[`Moon should return 0.5 if the "origin" parameter is not provided 1`] = `0.03596620967759451`;

exports[`Moon should return a MoonIlluminationData object when called with a number parameter 1`] = `
Object {
  "angle": 2.042516805424306,
  "fraction": 0.1878109225830057,
  "next": Object {
    "date": "2021-11-04T23:26:30.060Z",
    "firstQuarter": Object {
      "date": "2021-11-12T08:37:30.754Z",
      "value": 1636706250754.5,
    },
    "fullMoon": Object {
      "date": "2021-11-19T17:48:31.449Z",
      "value": 1637344111449,
    },
    "newMoon": Object {
      "date": "2021-11-04T23:26:30.060Z",
      "value": 1636068390060,
    },
    "thirdQuarter": Object {
      "date": "2021-11-27T02:59:32.143Z",
      "value": 1637981972143.5,
    },
    "type": "newMoon",
    "value": 1636068390060,
  },
  "phase": Object {
    "code": ":waning_crescent_moon:",
    "css": "wi-moon-wan-cres",
    "emoji": "ðŸŒ˜",
    "from": 0.783863193308711,
    "id": "waningCrescentMoon",
    "name": "Waning Crescent",
    "to": 0.966136806691289,
    "weight": 6.3825,
  },
  "phaseValue": 0.8573237783716006,
}
`;

exports[`Moon should return an object with 'az', 'el', 'rng', and 'parallacticAngle' properties 1`] = `
Object {
  "az": 5.030114468719165,
  "el": -0.12433217612739415,
  "parallacticAngle": 0.8650448747993286,
  "rng": 380743.9840372972,
}
`;

exports[`Moon should return an object with 'az', 'el', 'rng', and 'parallacticAngle' properties 2`] = `
Object {
  "az": 4.034314328077605,
  "el": 0.19547892013212267,
  "parallacticAngle": 0.7236664732665937,
  "rng": 366252.754397278,
}
`;

exports[`Moon should return with 'rise', 'set', 'ye', 'alwaysUp', 'alwaysDown', 'highest' properties 1`] = `
Object {
  "alwaysDown": false,
  "alwaysUp": false,
  "highest": 2024-01-12T21:42:31.618Z,
  "rise": 2024-01-12T16:43:59.167Z,
  "set": 2024-01-13T02:41:04.069Z,
  "ye": null,
}
`;

exports[`Moon should return with 'rise', 'set', 'ye', 'alwaysUp', 'alwaysDown', and 'highest' properties 1`] = `
Object {
  "alwaysDown": false,
  "alwaysUp": false,
  "highest": 2024-01-12T21:42:31.618Z,
  "rise": 2024-01-12T16:43:59.167Z,
  "set": 2024-01-13T02:41:04.069Z,
  "ye": null,
}
`;
